url7b90c1010b_DastanIqbal_HelloZMQwithJava_tgz-pJ8-ClientJ8/src/zmq/StreamEngine.java

```diff
@@ -0,0 +1,623 @@
+                    error();
+                }
+
+                return;
+            }
+        }
+
+        //  If there are any data to write in write buffer, write as much as
+        //  possible to the socket. Note that amount of data to write can be
+        //  arbitratily large. However, we assume that underlying TCP layer has
+        //  limited transmission buffer and thus the actual number of bytes
+        //  written should be reasonably modest.
+        int nbytes = write (outbuf);
+
+        //  IO error has occurred. We stop waiting for output events.
+        //  The engine is not terminated until we detect input error;
+        //  this is necessary to prevent losing incomming messages.
+        if (nbytes == -1) {
+            io_object.reset_pollout (handle);
+
+            if (terminating)
+                terminate ();
+
+            return;
+        }
+
+        outsize -= nbytes;
+
+        //  If we are still handshaking and there are no data
+        //  to send, stop polling for output.
+        if (handshaking)
+            if (outsize == 0)
+                io_object.reset_pollout (handle);
+        
+        // when we use custom encoder, we might want to close after sending a response
+        if (outsize == 0) {
+            if (encoder != null && encoder.is_error ()) {
+                error();
+                return;
+            }
+            if (terminating)
+                terminate ();
+        }
+
+    }
+    
+
+// JDK6    @Override
+    public void connect_event() {
+        throw new UnsupportedOperationException();
+        
+    }
+
+// JDK6    @Override
+    public void accept_event() {
+        throw new UnsupportedOperationException();
+        
+    }
+
+// JDK6    @Override
+    public void timer_event(int id_) {
+        throw new UnsupportedOperationException();
+        
+    }
+
+    
+// JDK6    @Override
+    public void activate_out() {
+        io_object.set_pollout (handle);
+
+        //  Speculative write: The assumption is that at the moment new message
+        //  was sent by the user the socket is probably available for writing.
+        //  Thus we try to write the data to socket avoiding polling for POLLOUT.
+        //  Consequently, the latency should be better in request/reply scenarios.
+        out_event ();
+    }
+
+
+    
+// JDK6    @Override
+    public void activate_in ()
+    {
+        if (!io_enabled) {
+            //  There was an input error but the engine could not
+            //  be terminated (due to the stalled decoder).
+            //  Flush the pending message and terminate the engine now.
+            decoder.process_buffer (inbuf, 0);
+            assert (!decoder.stalled ());
+            session.flush ();
+            error ();
+            return;
+        }
+
+        io_object.set_pollin (handle);
+
+        //  Speculative read.
+        io_object.in_event ();
+    }
+    
+    private boolean handshake ()
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 16. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.